Set Title and start date and time for duplicate events using duplicate WFH

tyfranck opened the pull request:
For rolling forward Lectures from one year to the next we need to be able to change the Title and Start Date and Time of the duplicated mediapackage.

This PR is based on PR #3409 .

Fix 2 issues highlighted in the comments of the above PR
- Date changes not reflected
- issues with empty date

### Your pull request should…

* [x] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

tyfranck committed:
Set Title and start date and time for duplicate events.

For rolling forward Lectures from one year to the next we need to be able to change the Title and Start Date and Time of the duplicated mediapackage.

This PR is based on PR #3409

Fix 2 issues highlighted in the comments
- Date changes not reflected
- issues with empty date

gregorydlogan commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -227,7 +246,7 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
     SeriesInformation series = null;
     AccessControlList seriesAccessControl = null;
-    if (!seriesId.isEmpty()) {
+    if (!seriesId.isEmpty() && !seriesId.startsWith("${")) {
```
This doesn't check that the substitution terminates.  Is that expected?

tyfranck replied:
fixed

gregorydlogan reviewed (approved):
LGTM, aside from my comment.

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/duplicate-event-woh.md:
```diff
@@ -16,8 +16,8 @@ For each duplicated event the new media package ID is stored as a workflow prope
 |----------------------------------------|--------------------------------------------------------------------|-----------------------------------------------|
-|duplicate\_media\_package\_*number*\_id |`duplicate_media_package_1_id=e72f2265-472a-49ae-bc04-8301d94b4b1a` |Media package ID of the duplicated event       | \*
-|duplicate\_media\_package\_ids          |`duplicate_media_package_ids=e72f2265-472a-49ae-bc04-8301d94b4b1a, a32e2265-472a-49ae-bc04-8351d94b4b1c` | comma separated list of Media package IDs of the duplicated event |
+|duplicate_media_package_*number*_id     |`duplicate_media_package_1_id=e72f2265-472a-49ae-bc04-8301d94b4b1a` |Media package ID of the duplicated event       | \*
```
Why did you remove the escape character on the underscore character?

tyfranck replied:
No idea why Tobias removed the escape character but looking at it closing, I can see that escape characters are not really needed in the name. There's no need to escape underscore character in the name.  So I guess it's a bit of clean up.

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -227,7 +246,7 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
     SeriesInformation series = null;
     AccessControlList seriesAccessControl = null;
-    if (!seriesId.isEmpty()) {
+    if (!seriesId.isEmpty() && !seriesId.startsWith("${")) {
```
The `.startsWith("${")` shouldn't be in here. If that's a value, it should be used

tyfranck replied:
When a series is not provided by the user, `${seriesId}` in the workflow definition is not substituted. So without that check the series will literally be set as `${seriesId}` in such instance.

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
         // Clone the media package (without its elements)
-        newMp = copyMediaPackage(mediaPackage, series, newMpId, noSuffix, i + 1, copyNumberPrefix);
+        String useTitle;
+        if (title.isEmpty() || (title.startsWith("${") && title.endsWith("}"))) {
```
Same here, the `(title.startsWith("${") && title.endsWith("}"))` is weird

tyfranck replied:
Same answer as above

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
-        newMp = copyMediaPackage(mediaPackage, series, newMpId, noSuffix, i + 1, copyNumberPrefix);
+        String useTitle;
+        if (title.isEmpty() || (title.startsWith("${") && title.endsWith("}"))) {
+          useTitle = mediaPackage.getTitle();
```
This method is deprecated. It's not guaranteed to yield the correct title. Get metadata from the DC catalog

tyfranck replied:
OK. That method needs @Deprecated Annotation if it can't been used anymore.

tyfranck replied:
fixed

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
+        String time = new SimpleDateFormat("HH:mm:ss").format(mpDate);
+        if (!startDateString.isEmpty()) {
+          String [] dt = startDateString.trim().split("\\s+");
+          if (dt.length >= 2) {
```
If `length > 2` that's clearly an error and you shouldn't guess values but throw an error

tyfranck replied:
fixed

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
+              date = dt[0];
+            }
+            if (!dt[1].startsWith("${")) {
+              time = (dt[1].split(":").length == 2) ? dt[1] + ":00" : dt[1];
```
The documentation above states `HH:MM:SS` as the format, this allows `HH:MM:SS` and `HH:MM`.

tyfranck replied:
Documentation updated to mention both format

tyfranck replied:
fixed with code revamped

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
+        try {
+          mpDate = ADMIN_UI_DATE_FORMAT.parse(date + " " + time);
+        } catch (ParseException ex) {
+          logger.info("{} could not be parsed as Date", date + " " + time);
```
This should fail the operation instead of doing something unexpected

tyfranck replied:
fixed

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
+        } catch (ParseException ex) {
+          logger.info("{} could not be parsed as Date", date + " " + time);
+        }
+        logger.info("setting StartDate to {}", mpDate.toString());
```
```suggestion
        logger.info("setting StartDate to {}", mpDate);
```

tyfranck replied:
fixed

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
+          logger.info("{} could not be parsed as Date", date + " " + time);
+        }
+        logger.info("setting StartDate to {}", mpDate.toString());
+        startDate = mpDate;
```
Do we need both variables?

tyfranck replied:
fixed

lkiesow commented on etc/workflows/duplicate-event.xml:
```diff
@@ -12,18 +12,48 @@
+          <br/>
+          <div>
+            <label for="startDate" style="width: 150px;">StartDate (YYYY-MM-DD):</label>
+            <input id="startDate" name="startDate" type="text" style="width: 200px;" value=" " class="configField" />
```
Use [`<input type="date">`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date)

tyfranck replied:
Yeah, I used date as a type before but the trouble is you can't really set a default value as `value =""` (empty). 
So when the date is not set, a null value is sent for that property which in turn throws an exception.
Properties values are enforced in `org.opencastproject.assetmanager.impl.persistence` package by the following code:

```
  public Property toProperty() {
    final PropertyId id = new PropertyId(mediaPackageId, namespace, propertyName);
    if (stringValue != null) {
      return new Property(id, Value.mk(stringValue));
    } else if (dateValue != null) {
      return new Property(id, Value.mk(dateValue));
    } else if (longValue != null) {
      return new Property(id, Value.mk(longValue));
    } else if (boolValue != null) {
      return new Property(id, Value.mk(boolValue));
    } else {
      throw new RuntimeException("Bug. At least one of the value columns must be non null.");
    }
  }
```
Basically input of type date or time can't be used unless a default value is set. Or maybe you have some suggestions.

mliradelc replied:
In that case, where the time is null, I suggest using the current time and log a warning for the event.

tyfranck replied:
fixed, current datetime set default in UI.

lkiesow commented on etc/workflows/duplicate-event.xml:
```diff
@@ -12,18 +12,48 @@
+          </div>
+          <div>
+            <label for="startTime" style="width: 150px;">startTime (HH:MM:SS):</label>
+            <input id="startTime" name="startTime" type="text" style="width: 200px;" value=" " class="configField" />
```
…and [`type=time`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/time)

mliradelc replied:
Same as above answer

tyfranck replied:
fixed with code revamped.

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
         }
         // Clone the media package (without its elements)
-        newMp = copyMediaPackage(mediaPackage, series, newMpId, noSuffix, i + 1, copyNumberPrefix);
+        String useTitle;
+        if (title.isEmpty() || (title.startsWith("${") && title.endsWith("}"))) {
+          useTitle = mediaPackage.getTitle();
+        } else {
+          useTitle = title;
+        }
+        final String newTitle = noSuffix
+            ? useTitle
+            : String.format("%s (%s %d)", useTitle, copyNumberPrefix, i + 1);
```
I did not go through all the code but this seems a complex way of:

```suggestion
        if (title.isEmpty()) {
          title = mediaPackage.getTitle();
        }
        if (!noSuffix) {
          title = String.format("%s (%s %d)", title, copyNumberPrefix, i + 1);
        }
```

tyfranck replied:
code simplified :-)

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -516,11 +565,16 @@ private MediaPackage copyDublinCore(
     final DublinCoreCatalog destinationDublinCore = DublinCoreUtil.loadEpisodeDublinCore(workspace, source).get();
     destinationDublinCore.setIdentifier(null);
     destinationDublinCore.setURI(sourceDublinCore.getURI());
+    destinationDublinCore.set(DublinCore.PROPERTY_CREATED, OpencastMetadataCodec.encodeDate(creationDate,Precision.Second));
```
```suggestion
    destinationDublinCore.set(DublinCore.PROPERTY_CREATED, OpencastMetadataCodec.encodeDate(creationDate, Precision.Second));
```

mliradelc replied:
Missing space here

tyfranck replied:
fixed

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -516,11 +565,16 @@ private MediaPackage copyDublinCore(
     destinationDublinCore.setURI(sourceDublinCore.getURI());
+    destinationDublinCore.set(DublinCore.PROPERTY_CREATED, OpencastMetadataCodec.encodeDate(creationDate,Precision.Second));
     destinationDublinCore.set(DublinCore.PROPERTY_TITLE, destination.getTitle());
+    DublinCoreValue eventTime = EncodingSchemeUtils.encodePeriod(new DCMIPeriod(creationDate, creationDate), Precision.Second);
```
This essentially sets the event date to 0.
If we don't have a period, we shouldn't set it

tyfranck replied:
fixed

tyfranck committed:
Update to set Title and start date and time for duplicate events following review

- Terminate substitution `endsWith` added
- simplified complex codes
- checked for period in episode before update
- updated start date format in documentation

tyfranck commented:
I've addressed most of the comments given my time. @lkiesow  I commented the one I didn't address. I don't see them as stopper but I let you judge.  Any suggestions welcome.

lkiesow commented:
@mliradelc if you are happy with the current state, feel free to merge.
We start the release process right now but still need some time to prepare the release notes.

mliradelc reviewed (changes requested):
Hi @tyfranck I just reviewed your code, is good to merge.
 
There are some things that can be improved:
- Use the DC catalog instead of deprecated method
- When the time and date values are missing, use the actual time and date, then log a warning.
- And a missing space.

I added the comments that @lkiesow has pointed out.

tyfranck committed:
Use DC catalog and set default date

- UI defaults to current date and time
- Date format changed
- Warning message logged when date is not set.
- Documentation updated to reflect new changes

tyfranck commented:
Hi @mliradelc, 
Thanks for taking a look. With the latest update to this PR, I've addressed your points raised above.
1- Using DC to get mediapackage Title. 
2- Current date time is set as default in UI
3- Space removed.
4- Documentation updated accordingly to reflect new changes.

tyfranck committed:
Fix tests error related to adding title to configuration

tyfranck commented:
Any hint on why this is failing `Test configuration` run ?

mliradelc commented on etc/workflows/duplicate-event.xml:
```diff
@@ -35,6 +60,28 @@
+        $(function () {
+            $('input[type="datetime-local"]').setNow();
+        });
+        
```
Here is the error, you need to delete the trailing spaces, that's why is failing the test

`etc/workflows/duplicate-event.xml:84`

tyfranck replied:
Thanks. Updated now.

tyfranck committed:
Remove trailing space
